ostd/src/arch/x86/timer/apic.rs

```diff
@@ -1,76 +1,115 @@
-        x86::kernel::{
-            apic::{self, DivideConfig},
-            tsc::TSC_FREQ,
-        },
+        tsc_freq,
     },
     trap::{IrqLine, TrapFrame},
 };
 
-/// Initializes APIC with tsc deadline mode or periodic mode.
-/// Return the corresponding [`IrqLine`] for the System Timer.
-pub(super) fn init() -> IrqLine {
-    init_tsc_freq();
+/// Initializes APIC with TSC-deadline mode or periodic mode.
+///
+/// Return the corresponding [`IrqLine`] for the system timer.
+pub(super) fn init_bsp() -> IrqLine {
     if is_tsc_deadline_mode_supported() {
-        info!("[Timer]: Enable APIC TSC deadline mode.");
-        init_tsc_mode()
+        init_deadline_mode_config();
     } else {
-        info!("[Timer]: Enable APIC periodic mode.");
-        init_periodic_mode()
+        init_periodic_mode_config();
     }
+
+    let timer_irq = IrqLine::alloc().unwrap();
+    init_timer(&timer_irq);
+    timer_irq
 }
 
-pub(super) static APIC_TIMER_CALLBACK: Once<Arc<dyn Fn() + Sync + Send>> = Once::new();
+/// Initializes APIC timer on AP.
+///
+/// The caller should provide the [`IrqLine`] for the system timer.
+pub(super) fn init_ap(timer_irq: &IrqLine) {
+    init_timer(timer_irq);
+}
+
+/// A callback that needs to be called on timer interrupt.
+pub(super) fn timer_callback() {
+    use x86::msr::{wrmsr, IA32_TSC_DEADLINE};
+
+    match CONFIG.get().expect("ACPI timer config is not initialized") {
+        Config::DeadlineMode { tsc_interval } => {
+            let tsc_value = unsafe { _rdtsc() };
+            let next_tsc_value = tsc_interval + tsc_value;
+            unsafe { wrmsr(IA32_TSC_DEADLINE, next_tsc_value) };
+        }
+        Config::PeriodicMode { .. } => {}
+    }
+}
 
 /// Determines if the current system supports tsc_deadline mode APIC timer
 fn is_tsc_deadline_mode_supported() -> bool {
+    use x86::cpuid::cpuid;
+
     const TSC_DEADLINE_MODE_SUPPORT: u32 = 1 << 24;
     let cpuid = cpuid!(1);
     (cpuid.ecx & TSC_DEADLINE_MODE_SUPPORT) > 0
 }
 
-fn init_tsc_mode() -> IrqLine {
-    let timer_irq = IrqLine::alloc().unwrap();
+fn init_timer(timer_irq: &IrqLine) {
+    match CONFIG.get().expect("ACPI timer config is not initialized") {
+        Config::DeadlineMode { .. } => {
+            init_deadline_mode(timer_irq);
+        }
+        Config::PeriodicMode { .. } => {
+            init_periodic_mode(timer_irq);
+        }
+    }
+}
+
+fn init_deadline_mode(timer_irq: &IrqLine) {
     // Enable tsc deadline mode
     apic::with_borrow(|apic| {
         apic.set_lvt_timer(timer_irq.num() as u64 | (1 << 18));
     });
-    let tsc_step = TSC_FREQ.load(Ordering::Relaxed) / TIMER_FREQ;
 
-    let callback = move || unsafe {
-        let tsc_value = _rdtsc();
-        let next_tsc_value = tsc_step + tsc_value;
-        wrmsr(IA32_TSC_DEADLINE, next_tsc_value);
-    };
+    timer_callback();
+}
 
-    callback.call(());
-    APIC_TIMER_CALLBACK.call_once(|| Arc::new(callback));
+fn init_periodic_mode(timer_irq: &IrqLine) {
```

Comment:
Let the caller pass `init_count` so we don't have to check the config and panic if it's unexpected?
```rust
        let Some(Config::PeriodicMode { init_count }) = CONFIG.get() else {
            panic!("APIC timer state not initialized as periodic mode");
       }
```